// @cdklabs/generative-ai-cdk-constructs/lib/patterns/gen-ai/aws-rag-appsync-stepfn-kendra/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceEndpointTypeEnum = exports.EndpointTypes = void 0;
var EndpointTypes;
(function (EndpointTypes) {
    EndpointTypes["GATEWAY"] = "Gateway";
    EndpointTypes["INTERFACE"] = "Interface";
})(EndpointTypes || (exports.EndpointTypes = EndpointTypes = {}));
var ServiceEndpointTypeEnum;
(function (ServiceEndpointTypeEnum) {
    ServiceEndpointTypeEnum["DYNAMODB"] = "DDB";
    ServiceEndpointTypeEnum["ECR_API"] = "ECR_API";
    ServiceEndpointTypeEnum["ECR_DKR"] = "ECR_DKR";
    ServiceEndpointTypeEnum["EVENTS"] = "CLOUDWATCH_EVENTS";
    ServiceEndpointTypeEnum["KENDRA"] = "KENDRA";
    ServiceEndpointTypeEnum["KINESIS_FIREHOSE"] = "KINESIS_FIREHOSE";
    ServiceEndpointTypeEnum["KINESIS_STREAMS"] = "KINESIS_STREAMS";
    ServiceEndpointTypeEnum["S3"] = "S3";
    ServiceEndpointTypeEnum["SAGEMAKER_RUNTIME"] = "SAGEMAKER_RUNTIME";
    ServiceEndpointTypeEnum["SECRETS_MANAGER"] = "SECRETS_MANAGER";
    ServiceEndpointTypeEnum["SNS"] = "SNS";
    ServiceEndpointTypeEnum["SQS"] = "SQS";
    ServiceEndpointTypeEnum["SSM"] = "SSM";
    ServiceEndpointTypeEnum["STEP_FUNCTIONS"] = "STEP_FUNCTIONS";
})(ServiceEndpointTypeEnum || (exports.ServiceEndpointTypeEnum = ServiceEndpointTypeEnum = {}));
// interface ServiceEndpointTypeEnum {
//   DYNAMODB: 'DDB';
//   ECR_API: 'ECR_API';
//   ECR_DKR: 'ECR_DKR';
//   EVENTS: 'CLOUDWATCH_EVENTS';
//   KENDRA: 'KENDRA';
//   KINESIS_FIREHOSE: 'KINESIS_FIREHOSE';
//   KINESIS_STREAMS: 'KINESIS_STREAMS';
//   S3: 'S3';
//   SAGEMAKER_RUNTIME: 'SAGEMAKER_RUNTIME';
//   SECRETS_MANAGER: 'SECRETS_MANAGER';
//   SNS: 'SNS';
//   SQS: 'SQS';
//   SSM: 'SSM';
//   STEP_FUNCTIONS: 'STEP_FUNCTIONS';
// }
