// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/topic-list.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Topic = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constructs_1 = require("constructs");
class Topic extends constructs_1.Construct {
    constructor(scope, id) {
        super(scope, id);
        this.topicConfigList = [];
    }
    financialAdviceTopic(props) {
        this.createTopic({
            name: props?.name ?? 'Financial_Advice',
            definition: props?.definition ?? 'Discussions that involve providing guidance, recommendations, or suggestions related to managing, investing, or handling finances, investments, or assets.',
            examples: props?.examples ?? [
                'Can you suggest some good stocks to invest in right now?',
                "What's the best way to save for retirement?",
                'Should I put my money in a high-risk investment?',
                'How can I maximize my returns on investments?',
                'Is it a good time to buy real estate?',
            ],
            type: props?.type ?? 'DENY',
        });
    }
    politicalAdviceTopic(props) {
        this.createTopic({
            name: props?.name ?? 'Political_Advice',
            definition: props?.definition ?? 'Conversations that express views, opinions, or endorsements related to political parties, political ideologies, elections, or political figures.',
            examples: props?.examples ?? [
                "What's your stance on the current government policies?",
                'Do you support party X or Y in the upcoming election?',
                'Can you tell me which political party is better?',
                'What do you think about the new policy introduced by the president?',
                'Should I vote for this candidate?',
            ],
            type: props?.type ?? 'DENY',
        });
    }
    medicalAdvice(props) {
        this.createTopic({
            name: props?.name ?? 'Medical_Advice',
            definition: props?.definition ?? 'Providing recommendations, diagnosis, treatment options, or guidance on medical conditions, symptoms, medications, or health-related issues.',
            examples: props?.examples ?? [
                'What should I do to treat a persistent cough?',
                'Can you recommend some medication for my headache?',
                'What are the symptoms of a specific medical condition?',
                'Is this drug effective for treating my illness?',
                "Do I need to see a doctor for this pain I'm experiencing?",
            ],
            type: props?.type ?? 'DENY',
        });
    }
    inappropriateContent(props) {
        this.createTopic({
            name: props?.name ?? 'Inappropriate_Content',
            definition: props?.definition ?? 'Any discussions or references that include hate speech, discriminatory remarks, sexual content, or explicit language.',
            examples: props?.examples ?? [
                'Why are people from X country so [discriminatory remark]?',
                'Can you tell me a dirty joke?',
                '[Use of explicit language]',
                'This service is as bad as [hate speech].',
                'Do you have any adult content or products?',
            ],
            type: props?.type ?? 'DENY',
        });
    }
    legalAdvice(props) {
        this.createTopic({
            name: props?.name ?? 'Legal_Advice',
            definition: props?.definition ?? 'Offering guidance or suggestions on legal matters, legal actions, interpretation of laws, or legal rights and responsibilities.',
            examples: props?.examples ?? [
                'Can I sue someone for this?',
                'What are my legal rights in this situation?',
                'Is this action against the law?',
                'What should I do to file a legal complaint?',
                'Can you explain this law to me?',
            ],
            type: props?.type ?? 'DENY',
        });
    }
    createTopic(props) {
        const topicConfig = {
            definition: props.definition || '',
            name: props.name || '',
            type: props.type || '',
            examples: props.examples || [],
        };
        this.topicConfigList.push(topicConfig);
    }
    topicConfigPropertyList() {
        return this.topicConfigList;
    }
}
exports.Topic = Topic;
_a = JSII_RTTI_SYMBOL_1;
Topic[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.Topic", version: "0.1.216" };
